pandas/core/aggregation.py

```diff
@@ -198,6 +268,79 @@ def maybe_mangle_lambdas(agg_spec: Any) -> Any:
+    func: Dict[str, List[Union[Callable, str]]],
+    columns: Tuple,
+    order: List[int],
+) -> Dict[Label, Series]:
+    """Internal function to reorder result if relabelling is True for
+    dataframe.agg, and return the reordered result in dict.
+
+    Parameters:
+    ----------
+    result: Result from aggregation
+    func: Dict of (column name, funcs)
+    columns: New columns name for relabelling
+    order: New order for relabelling
+
+    Examples:
+    ---------
+    >>> result = DataFrame({"A": [np.nan, 2, np.nan],
+    ...       "C": [6, np.nan, np.nan], "B": [np.nan, 4, 2.5]})  # doctest: +SKIP
+    >>> funcs = {"A": ["max"], "C": ["max"], "B": ["mean", "min"]}
+    >>> columns = ("foo", "aab", "bar", "dat")
+    >>> order = [0, 1, 2, 3]
+    >>> _relabel_result(result, func, columns, order)  # doctest: +SKIP
+    dict(A=Series([2.0, NaN, NaN, NaN], index=["foo", "aab", "bar", "dat"]),
+         C=Series([NaN, 6.0, NaN, NaN], index=["foo", "aab", "bar", "dat"]),
+         B=Series([NaN, NaN, 2.5, 4.0], index=["foo", "aab", "bar", "dat"]))
+    """
+    reordered_indexes = [
+        pair[0] for pair in sorted(zip(columns, order), key=lambda t: t[1])
+    ]
+    reordered_result_in_dict: Dict[Label, Series] = {}
+    idx = 0
+
+    reorder_mask = len(result.columns) > 1
+    for col, fun in func.items():
+        s = result[col].dropna()
+
+        # In the `_aggregate`, the callable names are obtained and used in `result`, and
+        # these names are ordered alphabetically. e.g.
+        #           C2   C1
+        # <lambda>   1  NaN
+        # amax     NaN  4.0
+        # max      NaN  4.0
+        # sum     18.0  6.0
+        # Therefore, the order of functions for each column could be shuffled
+        # accordingly so need to get the callable name if it is not parsed names, and
+        # reorder the aggregated result for each column.
+        # e.g. if df.agg(c1=("C2", sum), c2=("C2", lambda x: min(x))), correct order is
+        # [sum, <lambda>], but in `result`, it will be [<lambda>, sum], and we need to
+        # reorder so that aggregated values map to their functions regarding the order.
+
+        # However there is only one column being used for aggregation, not need to
+        # reorder since the index is not sorted, and keep as is in `funcs`, e.g.
+        #         A
+        # min   1.0
+        # mean  1.5
+        # mean  1.5
+        if reorder_mask:
+            fun = [
+                com.get_callable_name(f) if not isinstance(f, str) else f for f in fun
+            ]
+            col_idx_order = Index(s.index).get_indexer(fun)
+            s = s[col_idx_order]
+
+        # assign the new user-provided "named aggregation" as index names, and reindex
+        # it based on the whole user-provided names.
+        s.index = reordered_indexes[idx : idx + len(fun)]
+        reordered_result_in_dict[col] = s.reindex(columns, copy=False)
+        idx = idx + len(fun)
+    return reordered_result_in_dict
```

Comment:
Hi, @WillAyd @jreback 

I think this part brought in quite conversations, and i looked back at it and feel there is still something could improve, so i re-designed the implementation a bit, and added quite detailed comment to explain.

The issue of `result` is, after the `self._aggregate` and `mangle_lamdbas`, the index of results will be obtained from callable name (e.g. `sum` -> `'sum'`, or `np.max` -> `'amax'`), or kept as is for string case (e.g. `'sum'` -> `'sum'`), or a parsed lambda func (e.g. `lambda x: min(x)` -> `'<lambda>'`), and index will be sorted.

so, if the call to function is `df.agg(aa=(C2, sum), ab=(C1, max), ac=(C1, np.max), ad=(C2, lambda x: min(x)), ae=(C1, sum))`, the order of functions is `C1: [max, np.max, sum]` and `C2: [sum, lambda]`, but it will look like something below:
```
             C2   C1
<lambda>      1  NaN
amax        NaN  4.0
max         NaN  4.0
sum        18.0  6.0
```
Therefore, need to find the func name in index, and its position, and do a reordering for `s`, which you see in `reorder_mask` clause. And afterwards, we need to replace those names with the names users provided in the call, e.g. `aa, ab, ae` etc

The reason to have the mask is, if the result is a DataFrame and there is only one column being used for aggregation, e.g. `df.agg(aa=(C1, sum), ab=(C1, mean))`, there will be no re-sort in index, and the order is the same as called in function.

I know this might bring in a lot of confusions, but i have tried many scenarios for corner cases, seems this solution works quite good, and overall this is to achieve 3 things: 
1. the index name of aggregated dataframe has the same order of appearance as in the call, 
2. the column names of aggregated dataframes has the same order of appearance of column names as in the call, 
3. the aggregated result mapping is correct. 

Many thanks for your precious time reviewing this PR and would love to hear your feedbacks! thanks in advance! @WillAyd @jreback

Reply:
emm, i _think_ CI failure is irrelevant